### ISSUE TITLE
Buildpack Upgrade Caused Apps to not be able to Restage

### ISSUE THREAD

**Issue by mkkeffeler** [cyberark/cloudfoundry-conjur-buildpack#89]
## Summary
We couldn't restage ANY app in our nonprod PCF instance after the team performed a regular buildpack upgrade to version cyberark-conjur-buildpack-v1.2.0.zip, they were on version cyberark-conjur-buildpack-v1.1.1.zip (side note, we need to say what gets packaged in here, otherwise tough to know changes between versions.)

## Reproducible
   * [ X] Always 
   
## Version/Tag number
1.1 buildpack and 1.2 buildpack

## Environment setup
in a container, PCF environment (version 2.9.14 of TAS). Using on-prem instance. the app in question also used the java buildpack, which is at version 4.32.1 (java buildpack offline). the app was springboot, 2.1.0.RELEASE.

## Additional Information
We could really use some help.

**Comment by mkkeffeler** [comment]
even more weird, that same secrets.yml I referenced as working still works with summon on a traditional VM. we need to have the same thing work on summon as here, when we can. this doesn't really make much sense that dashes or underscores cause a breaking chang.

**Comment by izgeri** [comment]
Hey @mkkeffeler - FYI I've revised your comments

We are actively preparing for a new release that should address the issue you're reporting - I can keep you updated on timelines on this issue, but we have build failures we are addressing before we can release. As soon as those are fixed, we'll be releasing a new version.

I'm sorry this is causing problems for you - I hope to be in touch within the next week or so with good news.

**Comment by mkkeffeler** [comment]
I wish we could have left a working example of what worked and what broke, so everyone can be aware. 


Can we understand why any release was introduced with this regression? I'd expect build failures to stop a release, not permit a broken release. This comment doesn't help me relay back to the company what we are doing to prevent issues from arising in the future?

**Comment by izgeri** [comment]
@mkkeffeler just to be clear, we did not release a version with build failures - and new build failures are currently stopping a release.

The issue you've identified is a new issue, and we are working diligently to address it.

I hope to have more good news soon, I greatly appreciate your patience.

**Comment by mkkeffeler** [comment]
Sure, in the meantime, can you share the place where breaking changes are documented? we can't find it, and don't see an issue discussing it. If this was a breaking change and it was documented, then that would simplify this.

**Comment by izgeri** [comment]
@mkkeffeler I'm looking at the project [changelog](https://github.com/cyberark/cloudfoundry-conjur-buildpack/blob/master/CHANGELOG.md) between 2.1.1 and 2.1.4 ([git diff](https://github.com/cyberark/cloudfoundry-conjur-buildpack/compare/v2.1.1...v2.1.4)) and based on that it's not clear to me what would cause the breaking change either. But I appreciate your feedback - we'll definitely work to reproduce the issue you experienced before this next release, and update our test automation and docs as appropriate to ensure other users don't hit this same issue.

**Comment by izgeri** [comment]
@mkkeffeler just wanted to update that we released a [new version](https://github.com/cyberark/cloudfoundry-conjur-buildpack/releases/tag/v2.1.5) of the buildpack today that should provide a **much** better UX for handling unsupported environment variable names. You can follow the [install instructions](https://github.com/cyberark/cloudfoundry-conjur-buildpack/#installing-the-conjur-buildpack) to upgrade your buildpack version.

I also wanted to share that we reviewed the breaking change you reported, and it looks like it's been a part of the buildpack since the v2.0.0 release, which was part of the v1.0.0 tile. This release converted the buildpack from one that used [meta-buildpack](https://github.com/cf-platform-eng/meta-buildpack) for invocation to a [supply buildpack](https://docs.cloudfoundry.org/buildpacks/understand-buildpacks.html#supply-script), and as indicated by the major version bumps represented a breaking change.

You made a good point in your report that it can be hard to determine the buildpack version - I filed #91 so that we can make an update that will help with that. Please add additional info there if you have any thoughts / ideas.

I hope this helps to clear things up, but if you have further questions about this project please feel free to open another GitHub issue. If you have questions about your specific deployment and have access to CyberArk support, you may also open an official support case. We really appreciate your continued feedback!

**Comment by mkkeffeler** [comment]
So are we saying that secrets with an underscore aren't valid? That was the original issue, secrets with underscore in the environment variable name stopped working.

If so, the readme needs a overhauling to stop using examples with underscores.

**Comment by izgeri** [comment]
@mkkeffeler in the example you shared, the app was unable to deploy with env var: `MC_oauth2customerserviceclient-secret` but it worked when you changed it to `MCoauth2customerserviceclientsecret`.

The key was not the underscore, it was the dash you removed. We document our environment variable requirements [here](https://github.com/cyberark/cloudfoundry-conjur-buildpack/blob/master/README.md#create-a-secretsyml-file), and in particular:
>  environment variable names included in the secret.yml file must be valid shell variable names. In particular, they may contain upper or lowercase letters, numbers, and underscores only.

I hope this helps. I'll check as to whether I can add back your original issue report details today in case that will help us in this conversation.
